File: lib/solvers/UselessViaRemovalSolver/can-section-move-to-layer.ts
import { segmentToBoxMinDistance } from "@tscircuit/math-utils"
import type { ConnectivityMap } from "circuit-json-to-connectivity-map"
import type { HighDensityRouteSpatialIndex } from "lib/data-structures/HighDensityRouteSpatialIndex"
import type { ObstacleSpatialHashIndex } from "lib/data-structures/ObstacleTree"
import type { HighDensityRoute } from "lib/types/high-density-types"
import type { RouteSection } from "./route-section"

export const canSectionMoveToLayer = ({
  currentSection,
  targetZ,
  route,
  hdRouteSHI,
  obstacleSHI,
  connMap,
  defaultTraceThickness,
  obstacleMargin,
  traceMargin,
  shouldCheckStaticGeometryForSegment,
}: {
  currentSection: RouteSection
  targetZ: number
  route: HighDensityRoute
  hdRouteSHI: Pick<
    HighDensityRouteSpatialIndex,
    "getConflictingRoutesForSegment"
  >
  obstacleSHI: ObstacleSpatialHashIndex
  connMap: ConnectivityMap
  defaultTraceThickness: number
  obstacleMargin: number
  traceMargin: number
  shouldCheckStaticGeometryForSegment?: (
    start: RouteSection["points"][number],
    end: RouteSection["points"][number],
  ) => boolean
}): boolean => {
  const currentTraceThickness = route.traceThickness ?? defaultTraceThickness
  const routeIds = [route.connectionName, route.rootConnectionName].filter(
    (id): id is string => id !== undefined,
  )

  for (let i = 0; i < currentSection.points.length - 1; i++) {
    const A = { ...currentSection.points[i], z: targetZ }
    const B = { ...currentSection.points[i + 1], z: targetZ }
    const conflictingRoutes = hdRouteSHI.getConflictingRoutesForSegment(
      A,
      B,
      currentTraceThickness / 2 + traceMargin,
    )

    for (const { conflictingRoute } of conflictingRoutes) {
      const conflictingRouteIds = [
        conflictingRoute.connectionName,
        conflictingRoute.rootConnectionName,
      ].filter((id): id is string => id !== undefined)
      const conflictingRouteIsSameNet = routeIds.some((routeId) =>
        conflictingRouteIds.some(
          (conflictingRouteId) =>
            conflictingRouteId === routeId ||
            connMap.areIdsConnected(conflictingRouteId, routeId),
        ),
      )
      if (conflictingRouteIsSameNet) continue

      // The index already checks each trace and via with its own copper
      // radius. Rechecking its route-level distance as trace copper can
      // discard a wider via's collision, including on intermediate layers.
      return false
    }

    if (shouldCheckStaticGeometryForSegment?.(A, B) === false) continue

    const segmentBox = {
      centerX: (A.x + B.x) / 2,
      centerY: (A.y + B.y) / 2,
      width: Math.abs(A.x - B.x),
      height: Math.abs(A.y - B.y),
    }
    const searchMargin = currentTraceThickness / 2 + obstacleMargin
    const obstacles = obstacleSHI.searchArea(
      segmentBox.centerX,
      segmentBox.centerY,
      segmentBox.width + searchMargin * 2,
      segmentBox.height + searchMargin * 2,
    )

    for (const obstacle of obstacles) {
      // Same-net pads and copper should not block via removal collision checks.
      const obstacleIsSameNet = routeIds.some((routeId) =>
        obstacle.connectedTo.some(
          (connectedId) =>
            connectedId === routeId ||
            connMap.areIdsConnected(connectedId, routeId),
        ),
      )
      if (obstacleIsSameNet) continue

      if (obstacle.__zLayers?.includes(targetZ)) {
        const isAtObstacle =
          (Math.abs(A.x - obstacle.center.x) < 0.01 &&
            Math.abs(A.y - obstacle.center.y) < 0.01) ||
          (Math.abs(B.x - obstacle.center.x) < 0.01 &&
            Math.abs(B.y - obstacle.center.y) < 0.01)
        if (isAtObstacle) continue
      }

      const distToObstacle = segmentToBoxMinDistance(A, B, obstacle)
      if (distToObstacle < searchMargin) return false
    }
  }

  return true
}

File: rust/trace-simplification/src/useless_via_removal_solver/can_section_move_to_layer.rs
use crate::types::{ConnectivityMap, Point, RouteRef};
use crate::data_structures::{high_density_route_spatial_index::RouteClearanceIndex, obstacle_tree::ObstacleSpatialHashIndex};
use super::route_section::RouteSection;
use crate::math_utils::segment_to_box_min_distance;
use crate::types::Point2;

pub fn can_section_move_to_layer(
    section: &RouteSection, target_z: f64, route: &RouteRef,
    hd_routes: &dyn RouteClearanceIndex, obstacles: &ObstacleSpatialHashIndex,
    conn_map: &ConnectivityMap, default_thickness: f64, obstacle_margin: f64,
    trace_margin: Option<f64>, check_static: Option<&dyn Fn(&Point, &Point) -> bool>,
) -> Result<bool, String> {
    let route = route.borrow();
    let thickness = if route.metadata.get("traceThickness").is_none_or(|value| value.is_null()) {
        default_thickness
    } else { route.trace_thickness };
    let trace_margin = trace_margin.unwrap_or(0.0);
    let ids = [Some(route.connection_name.as_str()), route.root_connection_name.as_deref()];
    for pair in section.points.windows(2) {
        let start = pair[0].borrow();
        let end = pair[1].borrow();
        let a = Point { x: start.x, y: start.y, z: target_z };
        let b = Point { x: end.x, y: end.y, z: target_z };
        let conflicts = hd_routes.get_conflicting_routes_for_segment(&a, &b, thickness / 2.0 + trace_margin);
        for conflict in conflicts {
            let other = conflict.conflicting_route.borrow();
            let other_ids = [Some(other.connection_name.as_str()), other.root_connection_name.as_deref()];
            if ids.iter().flatten().any(|id| other_ids.iter().flatten().any(|other_id| other_id == id || conn_map.are_ids_connected(other_id, id))) { continue; }
            let other_thickness = if other.metadata.get("traceThickness").is_none_or(|value| value.is_null()) { default_thickness } else { other.trace_thickness };
            let radius = if trace_margin > 0.0 { crate::math_utils::max(other_thickness / 2.0, other.via_diameter / 2.0) } else { other_thickness / 2.0 };
            if conflict.distance < thickness / 2.0 + radius + trace_margin { return Ok(false); }
        }
        if check_static.is_some_and(|check| !check(&a, &b)) { continue; }
        let margin = thickness / 2.0 + obstacle_margin;
        for obstacle in obstacles.search_area((a.x + b.x) / 2.0, (a.y + b.y) / 2.0, (a.x - b.x).abs() + margin * 2.0, (a.y - b.y).abs() + margin * 2.0)? {
            let obstacle = obstacle.borrow();
            if ids.iter().flatten().any(|id| obstacle.connected_to.iter().any(|other| other == id || conn_map.are_ids_connected(other, id))) { continue; }
            if obstacle.z_layers.contains(&target_z) {
                let at_obstacle = ((a.x - obstacle.center.x).abs() < 0.01 && (a.y - obstacle.center.y).abs() < 0.01)
                    || ((b.x - obstacle.center.x).abs() < 0.01 && (b.y - obstacle.center.y).abs() < 0.01);
                if at_obstacle { continue; }
            }
            if segment_to_box_min_distance(Point2 { x: a.x, y: a.y }, Point2 { x: b.x, y: b.y }, obstacle.center, obstacle.width, obstacle.height) < margin { return Ok(false); }
        }
    }
    Ok(true)
}
